### ISSUE TITLE
Worse performance than that reported in the paper

### ISSUE THREAD

**Issue by tensor2numpy** [prdwb/orconvqa-release#3]
Hello,
    Thanks for your nice work. I loaded the pipline checkpoint and can get the same result as reported in the paper. But when I trained the full model in the concurrent learning phase, I got a worse performance than that reported in the paper. Specifically,  the checkpoint of the Retriever Pretraining phase and all the data used in my running are released by the author. I used the same NVIDIA TITAN X GPU and kept hyper paramaters same as claimed in the paper and github(except the file path, there is no change). I run the code for two times and the results are as follows.
**The first time:**
Dev
{
  "unfiltered_f1": 26.169438318852666,
  "f1": 26.16943831885273,
  "HEQ": 16.41399416909621,
  "DHEQ": 0.0,
  "yes/no": 74.81049562682216,
  "followup": 51.924198250728864,
  "unanswerable_acc": 55.48172757475083,
  "rerank_mrr": 0.5068367346938776,
  "rerank_recall": 0.5536443148688046,
  "retriever_mrr": 0.4195724003887269,
  "retriever_recall": 0.5536443148688046,
  "global_step": "47290"
}
Test
{
  "unfiltered_f1": 26.301154000365973,
  "f1": 27.536951886484935,
  "HEQ": 21.905144694533764,
  "DHEQ": 0.648508430609598,
  "yes/no": 78.7781350482315,
  "followup": 56.491157556270096,
  "unanswerable_acc": 51.30434782608695,
  "rerank_mrr": 0.31796804882426855,
  "rerank_recall": 0.321994854305032,
  "retriever_mrr": 0.22769999401663377,
  "retriever_recall": 0.321994854305032
}
I found that the best dev result was obtained in step 47290 and f1 in the test is 27.5, which is worse than 29.4(reported in the paper. I also log the result in step 45000 and the result is still not ideal.
Dev
{"unfiltered_f1_45000": 25.9404748955696,
  "f1_45000": 25.940474895569675,
  "HEQ_45000": 15.947521865889213,
  "DHEQ_45000": 0.0,
  "yes/no_45000": 74.81049562682216,
  "followup_45000": 51.924198250728864,
  "unanswerable_acc_45000": 55.3156146179402,
  "rerank_mrr_45000": 0.5034256559766764,
  "rerank_recall_45000": 0.5504373177842565,
  "retriever_mrr_45000": 0.4132410106899903,
  "retriever_recall_45000": 0.5504373177842565,
}
Test
{
  "unfiltered_f1": 26.16817036093454,
  "f1": 27.41704909409341,
  "HEQ": 21.784565916398712,
  "DHEQ": 0.38910505836575876,
  "yes/no": 78.7781350482315,
  "followup": 56.491157556270096,
  "unanswerable_acc": 52.52173913043478,
  "rerank_mrr": 0.31087775982767907,
  "rerank_recall": 0.31433614551546696,
  "retriever_mrr": 0.2242027164482738,
  "retriever_recall": 0.31433614551546696
}

**The second time:**
Dev
{
  "unfiltered_f1": 26.169438318852666,
  "f1": 26.16943831885273,
  "HEQ": 16.41399416909621,
  "DHEQ": 0.0,
  "yes/no": 74.81049562682216,
  "followup": 51.924198250728864,
  "unanswerable_acc": 55.48172757475083,
  "rerank_mrr": 0.5068367346938776,
  "rerank_recall": 0.5536443148688046,
  "retriever_mrr": 0.4195724003887269,
  "retriever_recall": 0.5536443148688046,
  "global_step": "47290"
}
Test
{
  "unfiltered_f1": 26.301154000365973,
  "f1": 27.536951886484935,
  "HEQ": 21.905144694533764,
  "DHEQ": 0.648508430609598,
  "yes/no": 78.7781350482315,
  "followup": 56.491157556270096,
  "unanswerable_acc": 51.30434782608695,
  "rerank_mrr": 0.31796804882426855,
  "rerank_recall": 0.321994854305032,
  "retriever_mrr": 0.22769999401663377,
  "retriever_recall": 0.321994854305032
}
I also increased the train eopchs from 3 to 6. And the result is as follows.
**trian epoches=6**
Dev
{
  "unfiltered_f1": 24.57448821069959,
  "f1": 24.57448821069961,
  "HEQ": 15.393586005830905,
  "DHEQ": 0.20408163265306123,
  "yes/no": 74.81049562682216,
  "followup": 51.924198250728864,
  "unanswerable_acc": 57.64119601328904,
  "rerank_mrr": 0.4923566569484937,
  "rerank_recall": 0.5530612244897959,
  "retriever_mrr": 0.4219339164237124,
  "retriever_recall": 0.5530612244897959
}
Test
{
  "unfiltered_f1": 26.05005371618406,
  "f1": 27.44249109634639,
  "HEQ": 22.688906752411576,
  "DHEQ": 0.7782101167315175,
  "yes/no": 78.7781350482315,
  "followup": 56.491157556270096,
  "unanswerable_acc": 60.26086956521739,
  "rerank_mrr": 0.29742116914976363,
  "rerank_recall": 0.3083378208580147,
  "retriever_mrr": 0.21914976365703343,
  "retriever_recall": 0.3083378208580147
}
**Analysis**
I can obtain the same result reported in the paper via loading the release dpipline checkpoint. But now I can not train the model to obtain the same reported result. 
1) Maybe there are some mistakes in the released version. Can you run the code code again to check the released code? 
2) If you can upload the dev results in each log step (5000, 10000, ......), it may help a lot.
3) If you need, I can attach the pacakges' versions list. Maybe it can help find the reason.

**Comment by prdwb** [comment]
Hi, thank you for the detailed analysis, and thank you for confirming our results can be obtained with the released ckpt.

We actually have verified that the results can be reproduced before releasing the code. The results log file you requested will be sent to your email. We used Python 3.7, Torch 1.2.0, Transformers 2.3.0, Faiss 1.6.1, pytrec_eval 0.5.

Are you using another gpu for retrieval as we do? I couldn't think of any specific reason at moment. I'll let know if I can think of anything.

**Comment by tensor2numpy** [comment]
Thanks for your quick reply! I believe this work is robust and it is very interesting.
I do use another gpu for retrieval(one titan x for retriever and another titan x for the model training). I used Python 3.7, Torch 1.2.0, Transformers 2.3.0, Faiss 1.6.3, pytrec_eval 0.5. There is only a little of difference about the Faiss(my is 1.6.3 and yours is 1.6.1). 
I compared my "all_results.json"(attached) with yours and it seems that the performance is worse than yours in each logging step. There must be something wrong. I summarized the possible changes as follows.
    1 I set the per_gpu_eval_batch_size=2 and yours is 4
    2 my Faiss version is 1.6.3 and yours is 1.6.1
    3 There are some warnings like"AGradient overflow.  Skipping step, loss scaler 0 reducing loss scale to xxx". It        seems that some warnings because the fp16. Did your running report this?
    4 I attached my hyper paramaters information. Besides the above changes, I think there is no difference. But I still attached it to make sure it.
    5 Do you save the log information, like the train_retriever.ipynb(the released now is a small sample not a full training). It will help a lot.
Lookforward to your reply!
